### PR TITLE
Devops build pipeline

### PR DESCRIPTION
**Pull request by ghost**
- Created front-end build pipeline.
- Added "pending" to failing unit tests in order to skip them until they get taken care of.
    https://jasmine.github.io/tutorials/your_first_suite#section-Pending_Specs
- Implemented ESLint.

### CODE REVIEW

**Review by SilentMrDave — APPROVED** [review]

File: frontend-build-pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# probably needed for running the Karma tests
+#
+# angular-devkit/build-angular
+# https://stackoverflow.com/questions/50333003/could-not-find-module-angular-devkit-build-angular
+- script: |
+    npm install --save-dev @angular-devkit/build-angular
+    npm install -g @angular/cli
+    npm install
+    ng build --prod
+  displayName: 'npm Install and Build'
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: cf [my review comments here](https://github.com/1058-Sophia-Gavrila-Forge/Portfolio-Frontend/pull/3#pullrequestreview-708776485). also:
- `@angular-devkit/build-angular` is already there in the package.json so re-adding it should be superfluous.
- installing the angular cli globally only to install it locally as well is a huge waste of time only to get to run `ng` instead of `npx ng` (i don't expect you to have any context for that)

so, this should be replaced by an `npm ci` step and an `npx ng build --prod` step.

File: frontend-build-pipeline.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+# continueOnError for now since unit tests fail
+- script: |
+    npx ng test --karma-config karma.conf.js --watch=false
+  displayName: 'Run Unit Tests'
+  continueOnError: True
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: another option besides continueOnError is edit the test files to make all the currently failing ones skip. nearly all modern test frameworks have easy ways to mark tests for being skipped. i recommend that approach since it puts the ball more in the devs' court, doesn't affect the pipeline, and doesnt let new test failures get through, but there's room for disagreement

File: frontend-build-pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+- script: |
+    npm install --save-dev @angular-devkit/build-angular
+    npm install -g @angular/cli
+    npm install
+    ng build --prod
+  displayName: 'npm Install and Build'
+
+# continueOnError for now since unit tests fail
+- script: |
+    npx ng test --karma-config karma.conf.js --watch=false
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: good work more or less figuring this part out. the `--karma-config` part is superfluous though since angular.json already tells it to read that file.

there are nicer ways to encapsulate running these commands with different arguments, both in the package.json as scripts, and in the angular.json as configurations, but that's more advanced and this is ok too.

File: frontend-build-pipeline.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  displayName: 'Run Unit Tests'
+  continueOnError: True
+
+- task: SonarCloudPrepare@1
+  inputs:
+    SonarCloud: 'SonarCloud'
+    organization: '1054-daniel-felleman-revature-swag-shop'
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: any values that will certainly change in the next project iteration should be parameters/variables.

File: frontend-build-pipeline.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    scannerMode: 'CLI'
+    configMode: 'manual'
+    cliProjectKey: '1054-Daniel-Felleman-Revature-Swag-Shop_RevatureShopFrontend'
+    cliProjectName: 'RevatureShopFrontend'
+    cliSources: '.'
+
+- task: SonarCloudAnalyze@1
```
Comment (escalonn) [inline frontend-build-pipeline.yml:80]: no sonar publish quality gate step? it's not necessary but it's a small nice to have

File: frontend-build-pipeline.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+- task: PublishPipelineArtifact@1
+  inputs:
+    targetPath: '$(Pipeline.Workspace)/s/dist/RevatureShopFrontend'
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: - don't hard-code the default agent path structure (the `/s` part) here, either use predefined variables better or just use relative paths, nothing wrong with relative paths since the default working directory is well-defined too.
- this step can be made a little more generic by publishing `dist/*` instead - then it will work for any project name without needing to be edited. (that folder path is defined in the angular.json and will be different for each angular app.)

File: frontend-build-pipeline.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    artifact: 'front-end-artifact'
+    publishLocation: 'pipeline'
```
Comment (escalonn) [inline frontend-build-pipeline.yml:90]: are you sure these are the right inputs? the docs say publishLocation is for the old PublishBuildArtifacts task, not for this one. the docs might be garbage.

File: karma.conf.js [hunk 8]
```diff
@@ -37,7 +37,7 @@ module.exports = function (config) {
     colors: true,
     logLevel: config.LOG_INFO,
     autoWatch: true,
-    browsers: ['Chrome'],
+    browsers: ['ChromeHeadless'],
```
Comment (escalonn) [inline karma.conf.js]: this changes the test experience for the devs in a negative way, since they don't see the browser environment of the tests, which brings several small benefits. fortunately it's possible to pass a `--browsers` argument to ng test to override this file, so you can revert the change to it.

File: package.json [hunk 9]
```diff
@@ -8,7 +8,6 @@
     "watch": "ng build --watch --configuration development",
     "test": "ng test"
   },
-  "private": true,
```
Comment (escalonn) [inline package.json:11]: does this cause any warnings about required fields to print when you run npm install?
if it's a package you'll publish, the CLI is stricter about what fields this file should have.

**Review by escalonn — CHANGES_REQUESTED** [review]

File: frontend-build-pipeline.yml (hunk 1 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: `@angular-devkit/build-angular` was there because when I was testing with a windows build agent the pipeline would fail without it. I have removed it now since we switched back to using Ubuntu.

Thanks for the information about `ng` and `npx ng`. I'll look into those. Changed `ng` to `npx ng`.

Got rid of `npm install` in favor of `npm ci`.

Comment (escalonn) [reply frontend-build-pipeline.yml]: that's strange. that that fixed it is probably a strange side-effect of the actual cause. in a case like this you can mark 'resolve conversation' after your reply.

File: frontend-build-pipeline.yml (hunk 2 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: From the research I've done about this, it seems that these unit tests are coming from auto-generated spec.ts files when the project was created. I looked up how to skip these but it seems that the answers say to just delete the spec.ts files. Other answers mentioned `ng new --skipTests` when creating the angular project to avoid generating these files. I'm not sure if the devs would want to keep these files or not, but for the time being I'll leave  `continueOnError: True` until I know more.

Comment (escalonn) [reply frontend-build-pipeline.yml]: the people who say "delete the files" are just code monkeys who don't understand what that means. the point of those generated template files is that the devs are supposed to implement actual testing in them once they implement actual behavior in the classes they test. sure they can delete them if they want to do testing with a different file structure or something, but just deleting them is only a "solution" if you're abandoning unit testing.

the tests are written with jasmine; the jasmine way to skip tests is by [marking them as "pending"](https://jasmine.github.io/tutorials/your_first_suite#section-Pending_Specs). i suggest the `pending()` approach so you can give a reason: that the test was failing before the iteration began.

File: frontend-build-pipeline.yml (hunk 1 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: Yeah I was trying out the Windows build agent on a hunch that Chrome vs ChromeHeadless could have been causing the unit tests to fail, but the results were the same. For some reason building on Ubuntu would fail without ChromeHeadless, but it worked fine as Chrome on the Windows build agent.

Sorry I forgot to address this when I resolved the conversation earlier: "cf my review comments here." I don't think I've heard the term cf before, or I just don't recall the acronym right now. What does cf mean?

File: frontend-build-pipeline.yml (hunk 2 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: I see, thanks for the link. I added the pending message to each unit test that was failing.

File: frontend-build-pipeline.yml (hunk 3 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: I may look into your suggestion about encapsulating the arguments. Removed `--karma-config`.

File: frontend-build-pipeline.yml (hunk 1 above)
Comment (escalonn) [reply frontend-build-pipeline.yml]: - the windows build agents have everything needed to run GUI apps on. the ubuntu build agents do not. that's just a thing that may not be documented and you learn on-the-job.
- https://en.wiktionary.org/wiki/cf.#English
- `ng build --prod` used to be the standard, but i just read they deprecated that option in favor of `ng build --configuration production` (or `-c production`, but i suggest long names for readability)

Comment (escalonn) [reply frontend-build-pipeline.yml]: update, that changed in angular 12 alongside some other things i didn't notice, with the net effect that `ng build` (or `npm run build`) by itself is fine, no mention of prod needed because it's the new default.

Comment (ghost) [reply frontend-build-pipeline.yml]: Alright thanks again for pointing things out. Changed `ng build --prod` to `ng build`.

File: frontend-build-pipeline.yml (hunk 4 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: Updated pipeline to use parameters.

File: frontend-build-pipeline.yml (hunk 5 above)
Comment (ghost) [reply frontend-build-pipeline.yml:80]: I plan to add this, I left it out before because we didn't have the quality gate configured yet.

File: frontend-build-pipeline.yml (hunk 6 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: Updated pipeline to use a better predefined variable, and published the whole dist folder to make it handle generic folder names inside of it. `$(Build.Repository.LocalPath)/dist`

dist/* was not working because the publish task does not support wildcards in the file path.

File: frontend-build-pipeline.yml (hunk 7 above)
Comment (ghost) [reply frontend-build-pipeline.yml:90]: It seems to work okay since we are able to see the artifacts on the pipeline runs. Those inputs were generated from the task tool. It goes to a 404 page when I click the "About this task" link so I guess the documentation on it is lacking.

File: karma.conf.js (hunk 8 above)
Comment (ghost) [reply karma.conf.js]: Thanks for pointing this out. Switched browser back to Chrome in karma.conf.js. Added `--browsers` argument to the unit test task.

File: package.json (hunk 9 above)
Comment (ghost) [reply package.json:11]: I don't think there were any about required fields, just a bunch about deprecated packages. I previously used a npm publish task for Azure Artifacts and it would not run unless this line was taken out. Since I'm not using that task anymore I put it back the way it was.

File: frontend-build-pipeline.yml [hunk 10]
```diff
@@ -6,8 +6,21 @@ trigger:
 pool:
   vmImage: ubuntu-latest
 
+parameters:
+- name: repoOrganization # the name of the repo's organization
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: this is the sonarcloud org, not the github repo's org. they are 1-to-1 for us, but even for us there is a case difference between them.

File: frontend-build-pipeline.yml [hunk 11]
```diff
@@ -16,30 +29,30 @@ steps:
 
 - script: |
     npm ci
-    npx ng build --prod
+    npx ng build
```
Comment (escalonn) [inline frontend-build-pipeline.yml:55]: let's do this in two separate steps, because each one takes a moderate amount of time and could fail for its own reasons, and because if somehow npm ci failed but ng build succeeded, we'd want to fail the build anyway since that would be weird.

File: frontend-build-pipeline.yml [hunk 12]
```diff
@@ -16,30 +29,30 @@ steps:
...
   displayName: 'npm Build'
 
 # skipped failing unit tests
 # https://jasmine.github.io/tutorials/your_first_suite#section-Pending_Specs
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: that comment would really fit more in the PR description than here, it doesn't really relate to the pipeline itself.

File: frontend-build-pipeline.yml [hunk 13]
```diff
@@ -16,30 +29,30 @@ steps:
...
 - script: |
-    npx ng test --watch=false
+    npx ng test --watch=false --browsers=ChromeHeadless
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: we want to be collecting code coverage eventually, doesn't have to be this PR but it should be soon

ng lint is also a good thing to include in an angular pipeline.

File: frontend-build-pipeline.yml [hunk 14]
```diff
@@ -16,30 +29,30 @@ steps:
...
     SonarCloud: 'SonarCloud'
-    organization: '1054-daniel-felleman-revature-swag-shop'
+    organization: ${{ parameters.repoOrganization }}
     scannerMode: 'CLI'
     configMode: 'manual'
-    cliProjectKey: '1054-Daniel-Felleman-Revature-Swag-Shop_RevatureShopFrontend'
-    cliProjectName: 'RevatureShopFrontend'
+    cliProjectKey: ${{ parameters.sonarCloudProjectKey }}
+    cliProjectName: $ {{ parameters.sonarCloudProjectName }}
     cliSources: '.'
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: include properties to identify the sources, coverage exclusions, and eventually the location of the coverage report file.

**Review by escalonn — CHANGES_REQUESTED** [review]

**Comment by escalonn** [comment]
request review via github's interface when ready for review

File: frontend-build-pipeline.yml (hunk 10 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: Yeah I got two mixed up earlier and didn't update the comment. I updated the comment to be more specific.

File: frontend-build-pipeline.yml (hunk 11 above)
Comment (ghost) [reply frontend-build-pipeline.yml:55]: Alright, separated into different steps.

File: frontend-build-pipeline.yml (hunk 12 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: Moved to PR description.

File: frontend-build-pipeline.yml (hunk 13 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: Implemented eslint and setup some default rules that all give warnings instead of errors to reduce friction with the devs. I let them know that once they get all the warnings cleared out they can switch those to errors if they want the pipeline to halt on linting issues. That way they don't need to worry about immediately having to fix all the linting issues.

I went back to using npm install instead of npm ci because the package-lock.json file no longer matches packages.json since I added eslint. I'm not sure if I should have uploaded an updated package-lock.json file instead, but it seems to make the pipeline a bit more robust in case someone forgets to update their package-lock.json. Let me know which way would be better.

File: frontend-build-pipeline.yml (hunk 14 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: Added parameters to include the mentioned sonar cloud options.

File: .eslintignore [hunk 15]
```diff
@@ -0,0 +1,10 @@
+# don't lint node_modules
+node_modules
+# don't lint build output
+dist
+# don't lint coverage output
+coverage
+# don't lint configuration
+karma.conf.js
+# don't lint spec.ts
+*.spec.ts
```
Comment (escalonn) [inline .eslintignore]: you generally do want to lint test files.

File: frontend-build-pipeline.yml [hunk 16]
```diff
@@ -16,27 +19,48 @@ parameters:
 - name: sonarCloudProjectName # sonar cloud's cli project name
   type: string
   default: 'RevatureShopFrontend'
+- name: sonarCloudExclude # files for sonar cloud to ignore
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: * to ignore specifically from coverage calculations, not from everything else.

File: frontend-build-pipeline.yml [hunk 17]
```diff
@@ -16,27 +19,48 @@ parameters:
...
   inputs:
     versionSpec: '12.14'
   displayName: 'Install Node.js'
 
 - script: |
-    npm ci
+    ls $(Build.Repository.LocalPath)
+
+- script: |
+    npm install
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: re. https://github.com/1054-Daniel-Felleman-Revature-Swag-Shop/RevatureShopFrontend/pull/1#discussion_r676759574

you run npm install when you intentionally want to update the package-lock file, such as when you've edited the package.json dependencies. so run that manually and commit the changes it makes to the package lock. then let the pipeline continue to do npm ci.
i see what you're saying about pipeline robustness, but that negates the point of the lock file. the lock file is there so that the app doesn't suddenly break because the latest version of some nested dependency (which an npm install with an empty node_modules is likely to grab) has a critical bug. it really is just up to whoever modifies the package.json to ensure the lockfile is also updated by running npm install.

File: frontend-build-pipeline.yml [hunk 18]
```diff
@@ -16,27 +19,48 @@ parameters:
...
-
 # Node.js is required to build Angular apps
 - task: NodeTool@0
   inputs:
     versionSpec: '12.14'
   displayName: 'Install Node.js'
 
 - script: |
-    npm ci
+    ls $(Build.Repository.LocalPath)
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: looks like leftover debug stuff

File: frontend-build-pipeline.yml [hunk 19]
```diff
@@ -16,27 +19,48 @@ parameters:
...
+
+- script: |
+    npm install
+  displayName: 'npm Install'
+  workingDirectory: $(Build.Repository.LocalPath)
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: you probably don't need these workingDirectory lines, isn't that already the default?
on the other hand, if you're parametrizing the other paths (like sonarCloudSources), that kind of suggests you want this pipeline to be able to handle the angular app being in a subdirectory (which is somewhat common) just by changing parameters. that would require a few changes, but it's not a big priority.

File: frontend-build-pipeline.yml [hunk 20]
```diff
@@ -16,27 +19,48 @@ parameters:
...
 - script: |
-    npx ng test --watch=false --browsers=ChromeHeadless
+    npx ng test --watch=false --browsers=ChromeHeadless --code-coverage
   displayName: 'Run Unit Tests'
+  workingDirectory: $(Build.Repository.LocalPath)
+
+# shopt -s globstar makes '**' search through folders of any depth
+- script: |
+    shopt -s globstar
+    npm run eslint src/**/*.ts
```
Comment (escalonn) [inline frontend-build-pipeline.yml]: this is relying on the shell to find the files, but eslint already has plenty of its own logic for finding files. just pass it the current directory, or `src` if you're leaving out `e2e`, and handle this via the .eslintrc and .eslintignore files. (also, not a big concern, but in some environments i don't know if i'd trust the shell to be able to handle a massive argument list if there were 1000s of files.)

**Review by escalonn — CHANGES_REQUESTED** [review]

File: frontend-build-pipeline.yml (hunk 20 above)
Comment (escalonn) [reply frontend-build-pipeline.yml]: (also, i think `npm run` only passes extra args to the underlying script when they're after ` -- `.)

Comment (escalonn) [reply frontend-build-pipeline.yml]: also also - the point of the builder package, and the edits to angular.json, is so that the `ng lint` commands will just run eslint for you. so your package.json script can just be "lint": "ng lint" (or, just run npx ng lint).
and devs can also just run "ng lint" to see the issues themselves, and possibly even "ng lint --fix" to auto fix some if the underlying implementation supports it, which i think it does.

also also also - i don't think your .eslintrc follows the practices recommended by the [@angular-eslint](https://github.com/angular-eslint/angular-eslint) docs, like using `overrides`.

File: frontend-build-pipeline.yml (hunk 17 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: Ah okay. Updated package-lock.json and switched back to using npm ci.

File: frontend-build-pipeline.yml (hunk 18 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: My bad, that shouldn't have been left in there. Cleaned up.

File: .eslintignore (hunk 15 above)
Comment (ghost) [reply .eslintignore]: My reasoning for ignoring *.spec.ts was because it seems like the linter doesn't understand it's a test file, or, at least the wrong rules are being applied to them. I may be able to use an override to only apply or exclude a certain subset of the rules to spec.ts files. I'll look into it.

Capture

File: frontend-build-pipeline.yml (hunk 16 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: I must have misunderstood when I made the parameter, updated comment and changed the name to sonarCloudCoverageExclude.

File: frontend-build-pipeline.yml (hunk 19 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: Yeah you're right it's not really necessary to have those lines in my case. That would be a nice feature to handle the app being in different places. For right now I'll just remove the unnecessary lines.

File: .eslintignore (hunk 15 above)
Comment (escalonn) [reply .eslintignore]: ah. that file is relying on being run in a context where jasmine provides some global variables, which eslint isn't automatically aware of. try this https://stackoverflow.com/questions/36952252/eslint-angular-and-jasmine-is-not-defined-no-undef

Comment (escalonn) [reply .eslintignore]: those answers do look old, though, so check the actual up-to-date docs with that as a guide

Comment (ghost) [reply .eslintignore]: Thanks for the info, that seemed to fix it.

Official docs say adding jasmine to the eslint environments adds all of the jasmine testing global variables for version 1.3 and 2.0.

Looking at the package.json, we're using jasmine 3.7 but I tested it, and it does not complain anymore about the spec.ts files so I guess the global variables haven't changed.

File: frontend-build-pipeline.yml (hunk 20 above)
Comment (ghost) [reply frontend-build-pipeline.yml]: I was confused earlier because I was getting a misleading error message before.
If I just tried running `npm run eslint src` it would say "No files matching the pattern "src" were found." This made me think that it either could not find the src folder even though it was in the correct directory, or, it would only work for files but not directories. Both of these initial theories were incorrect though. I eventually found this, which cleared up my understanding.

https://stackoverflow.com/a/42611291

Apparently eslint defaults to .js if you don't specify an extension, so it looked for .js files in src and couldn't find any, then said nothing matching src was found. Now that I've figured this out, I shouldn't need the shell expansion anymore.

I was able to do away with needing `npm run eslint`, now I use `npx eslint src --ext .ts` which I think is at least a step in the right direction. But I can't figure out the error that `npx ng lint` is giving me. 

> Schema validation failed with the following errors:  Data path "" must have required property 'lintFilePatterns'.

I'll follow up on the .eslintrc formatting soon.

Comment (ghost) [reply frontend-build-pipeline.yml]: Got `npx ng lint` working, converted configuration to .eslintrc.json and fixed formatting.

**Review by escalonn — APPROVED** [review]
